eventrouter/internal/queue_posix.c

```diff
@@ -0,0 +1,237 @@
+#include "queue_.h"
+
+#include <assert.h>
+#include <pthread.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <sys/errno.h>
+#include <sys/time.h>
+
+//==============================================================================
+// Type Definitions
+//==============================================================================
+
+typedef struct
+{
+    // Manage concurrency.
+    pthread_mutex_t m_mutex;
+    pthread_cond_t m_cond;
+    // Manage contents.
+    int m_idx;      //< The next index to read from.
+    size_t m_size;  //< The number of elements in the queue.
+    // Manage storage.
+    size_t m_capacity;    //< The maximum number of elements the queue can hold.
+    ErEvent_t* m_data[];  //< The space where the data is held.
+} Queue_t;
+
+//==============================================================================
+// Local Functions
+//==============================================================================
+
+static ErEvent_t* read(Queue_t* a_queue)
+{
+    ErEvent_t* result = a_queue->m_data[a_queue->m_idx];
+    a_queue->m_idx    = (a_queue->m_idx + 1) % a_queue->m_capacity;
+    a_queue->m_size -= 1;
+    return result;
+}
+
+static void write(Queue_t* a_queue, ErEvent_t* a_event)
+{
+    int write_idx = (a_queue->m_idx + a_queue->m_size) % a_queue->m_capacity;
+    a_queue->m_data[write_idx] = a_event;
+    a_queue->m_size += 1;
+}
+
+/// Returns a `struct timespec` corresponding to the time `a_ms` in the future.
+static struct timespec future(int64_t a_ms)
+{
+    struct timeval tv;
+    gettimeofday(&tv, NULL);
+    const int64_t nanos = (tv.tv_sec * 1e9) + (a_ms * 1e6) + (tv.tv_usec * 1e3);
+
+    struct timespec ts;
+    ts.tv_sec  = nanos / 1000000000;
+    ts.tv_nsec = nanos % 1000000000;
+    return ts;
+}
+
+//==============================================================================
+// Public Functions
+//==============================================================================
+
+ErQueue_t ErQueueNew(size_t a_capacity)
+{
+    Queue_t* const result =
+        malloc(sizeof(*result) + (sizeof(result->m_data[0]) * a_capacity));
+
+    pthread_mutex_init(&result->m_mutex, NULL);
+    pthread_cond_init(&result->m_cond, NULL);
+    result->m_idx      = 0;
+    result->m_size     = 0;
+    result->m_capacity = a_capacity;
+
+    return result;
+}
+
+void ErQueueFree(ErQueue_t a_queue)
+{
+    assert(a_queue != NULL);
+    Queue_t* q = a_queue;
+    pthread_cond_destroy(&q->m_cond);
+    pthread_mutex_destroy(&q->m_mutex);
+    free(q);
+}
+
+ErEvent_t* ErQueuePopFront(ErQueue_t a_queue)
+{
+    assert(a_queue != NULL);
+
+    Queue_t* q        = a_queue;
+    ErEvent_t* result = 0;
+
+    // Block until there is data in the queue.
+    while (1)
+    {
+        // Check the condition and while holding the mutex to avoid a race.
+        pthread_mutex_lock(&q->m_mutex);
```

Comment:
Can you move `pthread_mutex_lock(&q->m_mutex);` outside of `while(1)`? If it is possible, we can call `pthread_mutex_unlock(&q->m_mutex);` only once.

Reply:
Got it. makes sense.

Reply:
**CORRECTION:** @hamasaki-koji-yolabs I believe I am wrong, and your suggestion will make this code cleaner. Thank you for pointing this out.